### DISCUSSION TITLE
How to apply a filter with netconf GetConfig in Scrapligo?

### DISCUSSION THREAD

**Discussion by bruckynetworkstest** [scrapli/scrapligo#131] [Q&A] [answered]
Hi all, 

Myself again, stuck again, apologies. 

I am currently working on a project where I'm using the Scrapligo library to interact with a NETCONF server. I have been successful in retrieving data using the Get operation with a filter, as shown in the following code snippet:

```
filter := `<interfaces xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-interfaces-oper"/>`
r, err := d.Get(filter)
if err != nil {
    fmt.Printf("failed executing Get; error: %+v\n", err)
    return
}
if r.Failed != nil {
    fmt.Printf("response object indicates failure: %+v\n", r.Failed)
    return
}
fmt.Printf("Config result: %s", r.Result)
```

However, I am now trying to use a similar filter with the GetConfig operation to retrieve specific configuration data, but I can't find a way to apply the filter.

The function signature for GetConfig is as follows:

`func (d *Driver) GetConfig(source string, opts ...util.Option) (*response.NetconfResponse, error)`

From the documentation, it isn't clear how to apply a filter using this function, if it's possible at all.

Could you please guide me on how to use a filter with the GetConfig operation in Scrapligo, or confirm if this is not currently supported?

Thank you in advance for your assistance, Cheers.

**Comment by carlmontanari** [comment, marked as answer]
think you're prolly lookin for [this guy](https://github.com/scrapli/scrapligo/blob/main/driver/opoptions/netconf.go#L39); basically getconfig like normal + pass the filter and *should* do the trick :)

**Reply by bruckynetworkstest** [reply]
Thanks again, spot on as always. I also managed to work out changing to xpath filter type from this as well which i was looking for, so thanks again, saved another silly Q from myself.

**Reply by carlmontanari** [reply]
np! would you mind pasting an example here for anyone else who runs into this... obviously not the most well documented thing on my end 😬

**Reply by bruckynetworkstest** [reply]
Yes sure, happy to. I have built up a lot of scripts using scrapligo and netconf the last month, which If this helps in any way I can provide. These include (all for netconf), get operations with filtering and parsing specific values , pretty outputs to the terminal as well as write output as xml or json, working with tables to store the data, working with net/http to display to a web terminal, edit-config operations (create, delete, replace) mac address lookups, ip address lookups, ping sweeper and after tonight a lot more filtering with GetConfig :)

The below is a working script for netconf GetConfig applying a subtree filter to get the memory from a cisco ios-xe device using the native module

```
package main

import (
	"fmt"

	"github.com/scrapli/scrapligo/driver/netconf"
	"github.com/scrapli/scrapligo/driver/opoptions"
	"github.com/scrapli/scrapligo/driver/options"
)

func main() {
	d, err := netconf.NewDriver(
		"x.x.x.x",
		options.WithAuthNoStrictKey(),
		options.WithAuthUsername("xxxxx"),
		options.WithAuthPassword("xxxxx"),
		options.WithPort(830),
	)
	if err != nil {
		fmt.Printf("failed to create driver; error: %+v\n", err)

		return
	}

	err = d.Open()
	if err != nil {
		fmt.Printf("failed to open driver; error: %+v\n", err)

		return
	}

	defer d.Close()

	// Apply a subtree filter
	filter := `<native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native"><memory/></native>`
	r, err := d.GetConfig("running", opoptions.WithFilter(filter))
	if err != nil {
		fmt.Printf("failed executing GetConfig; error: %+v\n", err)

		return
	}
	if r.Failed != nil {
		fmt.Printf("response object indicates failure: %+v\n", r.Failed)

		return
	}

	fmt.Printf("Config result: %s", r.Result)
}
```

The below is a working script for netconf GetConfig applying a xpath filter to get the memory from a cisco ios-xe device using the native module

```
package main

import (
	"fmt"

	"github.com/scrapli/scrapligo/driver/netconf"
	"github.com/scrapli/scrapligo/driver/opoptions"
	"github.com/scrapli/scrapligo/driver/options"
)

func main() {
	d, err := netconf.NewDriver(
		"x.x.x.x",
		options.WithAuthNoStrictKey(),
		options.WithAuthUsername("xxxxx"),
		options.WithAuthPassword("xxxxx"),
		options.WithPort(830),
	)
	if err != nil {
		fmt.Printf("failed to create driver; error: %+v\n", err)

		return
	}

	err = d.Open()
	if err != nil {
		fmt.Printf("failed to open driver; error: %+v\n", err)

		return
	}

	defer d.Close()

	// Apply a xpath filter
	filter := `/native/memory`
	r, err := d.GetConfig("running", opoptions.WithFilterType("xpath"), opoptions.WithFilter(filter))

	if err != nil {
		fmt.Printf("failed executing GetConfig; error: %+v\n", err)

		return
	}
	if r.Failed != nil {
		fmt.Printf("response object indicates failure: %+v\n", r.Failed)

		return
	}

	fmt.Printf("Config result: %s", r.Result)
}
```

**Comment by carlmontanari** [comment]
you rock, thank you!!

> Yes sure, happy to. I have built up a lot of scripts using scrapligo and netconf the last month, which If this helps in any way I can provide.

If you are up for contributing these as examples that would be amazing! alternatively and a bit less effort just posting discussions and answering your own discussion with things would be great too! or none of the above, no pressure at all this is already great!

thanks!

**Reply by bruckynetworkstest** [reply]
OK cool, I'll work on getting some kind of structure together and adding some more useful comments and get these examples over here.
